### PR TITLE
Refactor(qwen3-vl) optimize position encoding interpolation

### PR DESCRIPTION
**Pull request by aaaandychen**
## Motivation

Sync Slime patches. Optimize fast_pos_embed_interpolate for Qwen3-VL by vectorizing core embedding lookups and interpolation math. This reduces GPU kernel launch overhead and improves throughput during batched inference.

related: [THUDM/slime#1316](https://github.com/THUDM/slime/issues/1316)

## Modifications
python/sglang/srt/models/qwen3_vl.py
Refactor fast_pos_embed_interpolate
- **Metadata Aggregation** Collect interpolation indices and weights within a loop to prepare for batch processing.Supplemented align_corners=False logic by referencing PyTorch's coordinate calculation in [UpSample.h](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/UpSample.h).
- **Batched Execution** Perform a single vectorized embedding lookup and weighted summation to minimize kernel overhead.
- **Reshaping & Compatibility** Reshape outputs to the required format.

test/srt/test_embed_interpolate_unittest.py
Refactor test input of fast_pos_embed_interpolate from `list` to `torch.Tensor` to align with existing methods.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) (`/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`) or contact authorized users to do so.
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/models/qwen3_vl.py [hunk 1]
```diff
@@ -397,28 +397,77 @@ def rot_pos_emb(
...
     def fast_pos_embed_interpolate(self, grid_thw):
-        patch_pos_embeds_permute = []
-        m_size = self.spatial_merge_size
+        grid_ts, grid_hs, grid_ws = grid_thw[:, 0], grid_thw[:, 1], grid_thw[:, 2]
+        device = self.pos_embed.weight.device
+
+        idx_list = [[] for _ in range(4)]
+        weight_list = [[] for _ in range(4)]
+
+        for t, h, w in zip(grid_ts, grid_hs, grid_ws):
```
Comment (zhaochenyang20) [inline python/sglang/srt/models/qwen3_vl.py]: Check that If grid_thw comes in as a GPU Tensor, iterating over it (for t, h, w in ...) and using h/w (scalar GPU tensors) as arguments for torch.linspace or shape calculations will trigger implicit CPU-GPU synchronization (equivalent to .item()) in every iteration. This kills the performance gain.

If so, please move everything to CPU first, then do the for loop.

File: python/sglang/srt/models/qwen3_vl.py [hunk 2]
```diff
@@ -397,28 +397,77 @@ def rot_pos_emb(
...
-            pos_embed = torch.nn.functional.interpolate(
-                embeds, size=(h, w), mode="bilinear", align_corners=self.align_corners
-            )
-            pos_embed = pos_embed.reshape(
-                -1,
-                h // self.spatial_merge_size,
-                self.spatial_merge_size,
-                w // self.spatial_merge_size,
-                self.spatial_merge_size,
+        pos_embeds = self.pos_embed(idx_tensor).to(device) * weight_tensor[:, :, None]
```
Comment (zhaochenyang20) [inline python/sglang/srt/models/qwen3_vl.py]: Same as Gemini, please remove this `to(device)`

File: python/sglang/srt/models/qwen3_vl.py [hunk 3]
```diff
@@ -397,28 +397,77 @@ def rot_pos_emb(
...
+        patch_pos_embeds = pos_embeds[0] + pos_embeds[1] + pos_embeds[2] + pos_embeds[3]
```
[the remark this answers is not in the document]
Comment (zhaochenyang20) [reply python/sglang/srt/models/qwen3_vl.py]: gemini is right. please use `torch.sum(pos_embeds, dim=0)`

File: python/sglang/srt/models/qwen3_vl.py [hunk 4]
```diff
@@ -397,28 +397,77 @@ def rot_pos_emb(
...
+            h_idxs_floor = h_idxs.int()
+            w_idxs_floor = w_idxs.int()
+            h_idxs_ceil = (h_idxs_floor + 1).clip(max=self.num_grid_per_side - 1)
+            w_idxs_ceil = (w_idxs_floor + 1).clip(max=self.num_grid_per_side - 1)
+
+            dh = h_idxs - h_idxs_floor
+            dw = w_idxs - w_idxs_floor
+
+            base_h = h_idxs_floor * self.num_grid_per_side
+            base_h_ceil = h_idxs_ceil * self.num_grid_per_side
```
Comment (zhaochenyang20) [inline python/sglang/srt/models/qwen3_vl.py]: The loop body is too dense (~40 lines) and repeats logic for h and w. Please extract the bilinear index/weight calculation into a private pure function.

File: python/sglang/srt/models/qwen3_vl.py [hunk 5]
```diff
@@ -397,28 +397,77 @@ def rot_pos_emb(
         return cos_combined, sin_combined
 
     def fast_pos_embed_interpolate(self, grid_thw):
```
Comment (zhaochenyang20) [inline python/sglang/srt/models/qwen3_vl.py:470]: Please add comment to this function.

File: python/sglang/srt/models/qwen3_vl.py [hunk 6]
```diff
@@ -397,28 +397,77 @@ def rot_pos_emb(
...
-        patch_pos_embeds_permute = []
-        m_size = self.spatial_merge_size
+        grid_ts, grid_hs, grid_ws = grid_thw[:, 0], grid_thw[:, 1], grid_thw[:, 2]
```
Comment (zhaochenyang20) [inline python/sglang/srt/models/qwen3_vl.py]: `grid_ts` seems an arbitrary naming. Could we name it more clearly?

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/qwen3_vl.py [hunk 7]
```diff
@@ -397,28 +397,77 @@ def rot_pos_emb(
...
+                h_idxs = torch.linspace(0, self.num_grid_per_side - 1, h, device=device)
+                w_idxs = torch.linspace(0, self.num_grid_per_side - 1, w, device=device)
+            else:
+                h_idxs = (torch.arange(h, device=device) + 0.5) * (
+                    self.num_grid_per_side / h
+                ) - 0.5
+                w_idxs = (torch.arange(w, device=device) + 0.5) * (
+                    self.num_grid_per_side / w
+                ) - 0.5
+                h_idxs = h_idxs.clamp(0, self.num_grid_per_side - 1)
```
Comment (zhaochenyang20) [inline python/sglang/srt/models/qwen3_vl.py]: in clamp and clip, only use one of them.

File: python/sglang/srt/models/qwen3_vl.py [hunk 8]
```diff
@@ -397,28 +397,77 @@ def rot_pos_emb(
...
+                w_idxs = w_idxs.clamp(0, self.num_grid_per_side - 1)
+
+            h_idxs_floor = h_idxs.int()
+            w_idxs_floor = w_idxs.int()
+            h_idxs_ceil = (h_idxs_floor + 1).clip(max=self.num_grid_per_side - 1)
```
Comment (zhaochenyang20) [inline python/sglang/srt/models/qwen3_vl.py]: in clamp and clip, only use one of them.

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]

**Comment by zhaochenyang20** [comment]
Great work also for the end2end profiling.

**Review by zhaochenyang20 — APPROVED** [review]
